packages/fork-choice/src/protoArray/protoArray.ts

```diff
@@ -214,6 +232,135 @@ export class ProtoArray {
     }
   }
 
+  /**
+   * Optimistic sync validate till validated latest hash, invalidate any decendant branch if invalidate till hash provided
+   * TODO: implementation:
+   * 1. verify is_merge_block if the mergeblock has not yet been validated
+   * 2. Throw critical error and exit if a block in finalized chain gets invalidated
+   */
+  validateLatestHash(latestValidExecHash: RootHex, invalidateTillBlockHash: RootHex | null): void {
+    // We can skip calling this if the block was successfully validated and is
+    // to be imported, as on_block would already propagate valid hash
+
+    // Look reverse because its highly likely node with latestValidExecHash is towards the
```

Comment:
is this latestValidExecHash unique in `this.nodes()`? If yes maybe we don't need to call `reverse()` for the performance reason?